### PR TITLE
feat: load all environment values from .env

### PR DESCRIPTION
**Pull request by anna-carroll**
anything that should be configurable by the operator, OR anything that will differ between testnet/mainnet --> now loaded via .env

### CODE REVIEW

**Comment by anna-carroll** [comment]
it's not perfect, some of these values would be generated programmatically by a "smarter" builder (e.g. `gas_limit` would be generated by simulating the block, `use_calldata` would be generated by checking whether blobs or calldata was cheaper, etc)

but probably better than hard-coding configured values && maintaining separate binaries for different chain environments

File: crates/builder/src/main.rs [hunk 1]
```diff
@@ -1,102 +1,55 @@
...
-        builder_rewards,
-        gas_limit: 30_000_000,
     };
 
     let (submit_channel, submit_jh) = submit.spawn();
 
     let (build_channel, build_jh) = build.spawn(submit_channel);
 
+    // server
     let server = serve_builder_with_span(build_channel, ([0, 0, 0, 0], 6969), span);
```
Comment (anna-carroll) [inline crates/builder/src/main.rs:53]: should the port be a configurable variable?

Comment (rswanson) [reply crates/builder/src/main.rs:53]: yes pls

Comment (anna-carroll) [reply crates/builder/src/main.rs]: https://github.com/init4tech/zenith-rs/pull/12/commits/c868567ab3668f61ecab9819374aed71876ed2f1

File: crates/builder/src/signer.rs [hunk 2]
```diff
@@ -23,6 +23,42 @@ impl LocalOrAws {
             Ok(LocalOrAws::Aws(signer))
         }
     }
+
+    /// Check if the key is present in the environment.
+    /// If so, load the signer (local privkey or AWS signer).
+    /// If not, return None.
+    pub async fn load_option(
```
Comment (prestwich) [inline crates/builder/src/signer.rs:30]: if we separate config loading from signer instantiation, we can delete all env access from this file, which feels a lot cleaner

File: crates/builder/src/tasks/block.rs [hunk 3]
```diff
@@ -84,6 +86,12 @@ pub struct BlockBuilder {
 }
 
 impl BlockBuilder {
+    pub fn new(config: &BuilderConfig) -> Self {
+        Self {
+            wait_secs: config.incoming_transactions_buffer,
```
Comment (prestwich) [inline crates/builder/src/tasks/block.rs:91]: ideally these would have the same name

**Review by prestwich — CHANGES_REQUESTED** [review]

File: crates/builder/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+pub fn load_address(key: &str) -> Result<Address, ConfigError> {
+    let address = env::var(key)?;
+    Address::from_str(&address).map_err(Into::into)
+}
+
+pub fn load_bool(key: &str) -> Result<bool, ConfigError> {
+    let val = env::var(key)?;
+    match val.as_str() {
+        "true" => Ok(true),
+        "false" => Ok(false),
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply crates/builder/src/config.rs:133]: sure https://github.com/init4tech/zenith-rs/pull/12/commits/02ac55b7c5d66c4f6308390caf1dcb5c75aafb77

File: crates/builder/src/config.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+pub enum ConfigError {
+    /// Error loading from environment variable
+    #[error("missing environment variable: {0}")]
+    Var(#[from] env::VarError),
+    /// Error parsing environment variable
+    #[error("failed to parse environment variable: {0}")]
+    Parse(#[from] num::ParseIntError),
+    /// Error parsing boolean environment variable
+    #[error("failed to parse boolean environment variable")]
+    ParseBool(),
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply crates/builder/src/config.rs]: https://github.com/init4tech/zenith-rs/pull/12/commits/55109b82dafcf5d6f649d4f62411ae8991f74f6e

File: crates/builder/src/signer.rs (hunk 2 above)
Comment (anna-carroll) [reply crates/builder/src/signer.rs]: yeah, that's much much nicer: https://github.com/init4tech/zenith-rs/pull/12/commits/45aafb676d943c14654c788f5be7becb0b873694

File: crates/builder/src/config.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        builder_signer: LocalOrAws::load(BUILDER_KEY, Some(host_chain)).await?,
+        incoming_transactions_buffer: load_u64(INCOMING_TRANSACTIONS_BUFFER)?,
+        block_confirmation_buffer: load_u64(BLOCK_CONFIRMATION_BUFFER)?,
+        submit_via_calldata: load_bool(SUBMIT_VIA_CALLDATA)?,
+        builder_rewards_address: load_address(BUILDER_REWARDS_ADDRESS)?,
+        rollup_block_gas_limit: load_u64(ROLLUP_BLOCK_GAS_LIMIT)?,
+    })
+}
+
+pub fn load_string(key: &str) -> Result<String, ConfigError> {
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply crates/builder/src/config.rs]: previously it was used in signer, but that's fixed now 😇 
https://github.com/init4tech/zenith-rs/pull/12/commits/45aafb676d943c14654c788f5be7becb0b873694

File: crates/builder/src/config.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+pub async fn load_builder_config() -> Result<BuilderConfig, ConfigError> {
+    let host_chain = load_u64(HOST_CHAIN_ID)?;
+    Ok(BuilderConfig {
+        host_chain_id: host_chain,
+        ru_chain_id: load_u64(RU_CHAIN_ID)?,
+        host_rpc_url: load_url(HOST_RPC_URL)?,
+        zenith_address: load_address(ZENITH_ADDRESS)?,
+        quincey_url: load_url(QUINCEY_URL)?,
+        builder_port: load_u16(BUILDER_PORT)?,
+        sequencer_signer: LocalOrAws::load_option(SEQUENCER_KEY, Some(host_chain)).await?,
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply crates/builder/src/config.rs]: this is indeed much nicer :~) 
https://github.com/init4tech/zenith-rs/pull/12/commits/45aafb676d943c14654c788f5be7becb0b873694

File: crates/builder/src/config.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+const BUILDER_KEY: &str = "BUILDER_KEY"; // AWS key ID (to use AWS signer) OR raw private key (to use local signer)
+const INCOMING_TRANSACTIONS_BUFFER: &str = "INCOMING_TRANSACTIONS_BUFFER";
+const BLOCK_CONFIRMATION_BUFFER: &str = "BLOCK_CONFIRMATION_BUFFER";
+const SUBMIT_VIA_CALLDATA: &str = "SUBMIT_VIA_CALLDATA";
+const BUILDER_REWARDS_ADDRESS: &str = "BUILDER_REWARDS_ADDRESS";
+const ROLLUP_BLOCK_GAS_LIMIT: &str = "ROLLUP_BLOCK_GAS_LIMIT";
+
+/// Configuration for a builder running a specific rollup on a specific host
+/// chain.
+pub struct BuilderConfig {
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply crates/builder/src/config.rs:23]: this was difficult while the Signer was still part of the struct but that's fixed now! done: https://github.com/init4tech/zenith-rs/pull/12/commits/45aafb676d943c14654c788f5be7becb0b873694

File: crates/builder/src/signer.rs [hunk 9]
```diff
@@ -16,15 +16,34 @@ pub enum LocalOrAws {
...
         }
     }
 }
 
+/// Load the wallet from environment variables.
+///
+/// # Panics
+///
+/// Panics if the env var contents is not a valid secp256k1 private key.
+pub fn wallet(private_key: &str) -> Result<LocalWallet, ConfigError> {
```
Comment (prestwich) [inline crates/builder/src/signer.rs:33]: these 2 free functions should be in the `impl LocalOrAws` block

File: crates/builder/src/config.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    AwsSigner(#[from] AwsSignerError),
+    /// Error parsing hex
+    #[error("failed to parse hex: {0}")]
+    Hex(#[from] hex::FromHexError),
+    /// Error loading the private key
+    #[error("failed to load private key: {0}")]
+    Wallet(#[from] alloy_signer_wallet::WalletError),
+}
+
+pub fn load_builder_config() -> Result<BuilderConfig, ConfigError> {
```
Comment (prestwich) [inline crates/builder/src/config.rs:78]: should be as follows:

```
impl BuilderConfig { 
    pub fn load_from_env() -> Result<Self, ConfigError> { 
        ...
    }
}
```

File: crates/builder/src/main.rs [hunk 11]
```diff
@@ -1,103 +1,60 @@
...
-    let config = ChainConfig::holesky(None, load_rpc()?, load_quincey()?, load_zenith_address()?);
-
-    // Load builder key from env
-    let wallet = LocalOrAws::load("BUILDER_KEY_ID", Some(config.host_chain_id)).await?;
-    let builder_rewards = wallet.address();
+    // load config from env
+    let config = load_builder_config()?;
 
+    // build provider from config
+    let builder_signer = LocalOrAws::load(&config.builder_key, Some(config.host_chain_id)).await?;
```
Comment (prestwich) [inline crates/builder/src/main.rs:26]: I would like this line to be 
`let builder_signer = config.connect_builder_signer().await?;`

which means a function 

```
impl BuilderConfig {
    /// Connect the configured builder signer 
    pub async fn connect_builder_signer(&self) -> Result<LocalOrAws, ConfigError>;
}

File: crates/builder/src/tasks/submit.rs [hunk 12]
```diff
@@ -198,7 +195,8 @@ where
 
         // If configured with a local signer, we use it. Otherwise, we ask
         // quincey (politely)
-        let signed = if let Some(signer) = &self.config.local_sequencer_signer {
+        let signed = if let Some(sequencer_key) = &self.config.sequencer_key {
```
Comment (prestwich) [inline crates/builder/src/tasks/submit.rs:198]: This reconnects the signer on every new inbound, which is probably not what we want. We should instead instantiate the signer in `spawn` and pass an `Option<LocalOrAws>` into `handle_inbound` as an argument

**Review by prestwich — CHANGES_REQUESTED** [review]

File: crates/builder/src/tasks/submit.rs (hunk 12 above)
Comment (prestwich) [reply crates/builder/src/tasks/submit.rs:198]: and like above, i'd like to see 

impl BuilderConfig {
    /// Connect the configured sequencer signer 
    pub async fn connect_sequencer_signer(&self) -> Option<Result<LocalOrAws, ConfigError>>;
}

File: crates/builder/src/signer.rs (hunk 9 above)
Comment (anna-carroll) [reply crates/builder/src/signer.rs]: https://github.com/init4tech/zenith-rs/pull/12/commits/dabe1c128b2bc6508209acff5cb4685636fcad94

File: crates/builder/src/tasks/submit.rs (hunk 12 above)
Comment (anna-carroll) [reply crates/builder/src/tasks/submit.rs]: https://github.com/init4tech/zenith-rs/pull/12/commits/8b0e0d49f02fe66f0cd3fb9dab3876731fc519cb

File: crates/builder/src/main.rs (hunk 11 above)
Comment (anna-carroll) [reply crates/builder/src/main.rs]: https://github.com/init4tech/zenith-rs/pull/12/commits/8b0e0d49f02fe66f0cd3fb9dab3876731fc519cb

**Review by prestwich — APPROVED** [review]

**Pull request merged**